### DISCUSSION TITLE
Infrastructure as Code

### DISCUSSION THREAD

**Discussion by lk-smit-ag** [zitadel/zitadel#2117] [Q&A] [answered]
Hello all,

we want to deploy a self-hosted instance of Zitadel on our Openshift-Cluster.
Because we have the "Infrastructure as Code" &  "Config as Code" approach, we want to get all required yaml-files into our private git repository (not github.com). Then we build it as a helm chart and integrate it with out CI/CD. But it looks like it can only be deployed with the "zitadelctl".

But i didn't figure out, how i can use the gitops mode with a private git server or any method to get the files for the zitadel deployment.
At the same time it would be nice, if we can change the namespace of the zitadel deployments/operators, can this be done?

So my questions are:

- Is it possible to get the yaml files for the zitadel deployment?
- Is it possible to change the namespace, where the system & zitadel is deployed?

**Comment by stebenz** [comment, marked as answer]
Hello,
Just to give a little bit of additional information, we specifically implemented zitadelctl so that no other tool is necessary to use ZITADEL in a cluster of your choice. As such, you only need zitadelctl to start the operator in your cluster, then the operator is self-reconciling and manages your ZITADEL.

About the general usage with Git, if you use GitOps you always have an "orbconfig"-file with 3 attributes, which provides zitadelctl with all that is needed to connect to the Git-repository. This file will get created with the "configure" command, and besides the URL to the repository and the masterkey, it contains the certificate used to connect to the repository, in Github and Gitlab they are called deploy-keys. ( I hope this clarifies it for you, otherwise don't hesitate to ask)

To answer the questions: 

- If you prepare a zitadel.yml in a Git-repository, which doesn't have to be on Github, you can use zitadelctl with the command "takeoff" and then you just need to read the resources applied to "caos-system", there are other ways as well, but I think this would be the easiest, as we have not currently planned to provide a helm chart. I would strongly recommend including the operator and not use the resources for ZITADEL directly.
- Currently, it is not possible to change the namespaces, we discussed the feature, but saw no need for ourselves, would it be necessary for you?

I hope I covered all the points if there are still questions don't hesitate to ask.

**Reply by lk-smit-ag** [reply]
Hello,
if i want to create a orbconfig with the command:
`zitadelctl --gitops configure --repourl "my_custom_git_repo" --masterkey "$(openssl rand -base64 21)"`
i will get redirected to github to confirm a login... Maybe i am doing something wrong?

I've applied the takeoff command to create the required objects and i could dump it. But the takeoff won't work for the caos-zitadel namespace in our environment.
The problems are currently:

**Certificate files for cockroachdb gets the wrong permisson**
This problems seems to be specific to OpenShift, because the mounted secrets gets a group level permission on the files. Which will be disallowed by cockroachdb by default.
Tried solutions:
- Mount the certificate on the file system and give right permissions => Operator overdeploys the configuration
- Set the environment COCKROACH_SKIP_KEY_PERMISSION_CHECK to true for the statefulset => Operator overdeploys the configuration

**Network Policies**
The next problem which i guess, is the connectivity from zitadel to the cockroachdb, because every pod connection will be only possible with network policies.. This can be deployed manually.

**Nonroot users**
The pod "ockroachdb-cluster-migration-ensure" wants to run as root user, which is disallowed in our environment.


So for me, the simplest way to have a custom cockroachdb (self configured & backuped) and a Deployment/StatefulSet/... for Zitadel. Or any way to configure each of it more in the deep (e.g. environment variables, ...). Maybe this is currently working, but i didn't find any documentation about it and with my "tryout" to set an env: on the spec of the database didn't work too.

Do you know any solution, how i can resolve the issues?

**Additional Information**
OpenShift Version: OKDv3.11.0+1cd89d4-542
Kubernetes Version: Kubernetesv1.11.0+d4cacc0

**Reply by stebenz** [reply]
Hello,

Ok, let's start with the zitadelctl-command, I just saw that it's necessary to add a "repokey"-parameter to the configure command so that it works for your example. I'm on it but the idea would be that a command like `zitadelctl --gitops configure --repourl "my_custom_git_repo" --masterkey "$(openssl rand -base64 21)" --repokey "$(cat repokey.ssh)` should cover what you need.

For cockroachDB, you can either for the moment being scale the operator to 0 replicas and make your changes, or I can provide you with the option to define an "ProvidedDatabase"-kind in the database.yml, so that you don't have to use the database-operator and just feed the information where cockroachDB is running. As a result you would have to deploy it yourself or use another operator to make it easier (https://catalog.redhat.com/software/operators/detail/5e9872712989e6a90307acd6). This "ProvidedDatabase" entails a little bit work on our part, so you would bear with me for a time.

Unfortunately, my knowledge for Openshift is limited, but if you would run cockroachDB yourself, you can deploy the necessary network policies right?

In case of the Nonroot users, this is indeed unfortunate, but I will try to change this as well.

**Reply by lk-smit-ag** [reply]
Hello, 
this sounds very nice. I will try out the gitops mode with the repokey.

Didn't thought about scaling the db-operator to zero, but i can give it a try!

I think that would be enough, to configure the cockroachdb for additional environment variables. So I can skip the Access Permission on Cert. Or we can provide another mount-path for the secrets, like a file storage.

And it would be very nice, if all containers run under a specific user or has set a non-root security context.

Thanks for your fast replies & help so far!

**Reply by adlerhurst** [reply]
hi @lk-smit-ag 

@stebenz is implementing nonroot users on [this branch](https://github.com/caos/zitadel/tree/container-nonroot).

Cheers

**Reply by adlerhurst** [reply]
@stebenz would be nice to get an update on this. thanks

**Reply by adlerhurst** [reply]
pull request is ready: https://github.com/caos/zitadel/pull/2137
